cmd/dependabot/internal/cmd/update.go

```diff
@@ -145,20 +80,117 @@ var updateCmd = &cobra.Command{
 	},
 }
 
-func readInputFile(file string) (*model.Input, []byte, error) {
+func extractInput(cmd *cobra.Command) (*model.Input, error) {
+	hasFile := file != ""
+	hasArguments := len(cmd.Flags().Args()) > 0
+	hasServer := inputServerPort != 0
+	hasStdin := doesStdinHaveData()
+
+	var count int
+	for _, b := range []bool{hasFile, hasArguments, hasServer, hasStdin} {
+		if b {
+			count++
+		}
+	}
+	if count > 1 {
+		return nil, errors.New("can only use one of: input file, arguments, server, or stdin")
+	}
+
+	if hasFile {
+		return readInputFile(file)
+	}
+
+	if hasArguments {
+		input, err := readArguments(cmd)
+		return input, err
+	}
+
+	if hasServer {
+		return server.Input(inputServerPort), nil
+	}
+
+	if hasStdin {
+		input, err := readStdin()
+		return input, err
+	}
+
+	return nil, fmt.Errorf("requires input as arguments, input file, or stdin")
+}
+
+func readStdin() (*model.Input, error) {
+	in := &bytes.Buffer{}
+	_, err := io.Copy(in, os.Stdin)
+	if err != nil {
+		return nil, err
+	}
+	data := in.Bytes()
+	input := &model.Input{}
+	if err = json.Unmarshal(data, &input); err != nil {
+		if err = yaml.Unmarshal(data, &input); err != nil {
+			return nil, fmt.Errorf("failed to decode input file: %w", err)
+		}
+	}
+	return nil, nil
```

Comment:
How does `input` get returned since this is `nil, nil`?

Reply:
Yeah that's a bug. I'm going to toss some tests around these once I figure out how I want to write them.

Reply:
I added a test for each input method. Had to do a bit of a refactor to get it to work as the cobra.Command has state, so had to create a constructor function.